mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,247 @@
+    mgr.set_constant("emirge_package_versions", output)
+
+
+def set_state(mgr: LogManager, state: ndarray) -> None:
+    """Update the state of all :class:`StateConsumer` of the log manager `mgr`."""
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                gd.quantity.set_state(state)
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self):
+        self.state = None
+
+    def set_state(self, state: ndarray) -> None:
+        """Set the state of the object."""
+        self.state = state
+
+
+class PhysicalQuantity(LogQuantity, StateConsumer):
+    """Logging support for physical quantities."""
+
+    def __init__(self, discr: Discretization, quantity: str, unit: str, op: str,
+                 name: str, rank_aggr: str = None):
+
+        LogQuantity.__init__(self, name, unit)
+        StateConsumer.__init__(self)
+
+        self.discr = discr
+
+        self.quantity = quantity
+
+        if rank_aggr is None:
+            self.rank_aggr = "norm_2"
+        else:
+            self.rank_aggr = rank_aggr
+
+        from functools import partial
+
+        if op == "min":
+            self._myop = partial(self.discr.nodal_min, "vol")
+        elif op == "max":
+            self._myop = partial(self.discr.nodal_max, "vol")
+        elif op == "sum":
+            self._myop = partial(self.discr.nodal_sum, "vol")
+        else:
+            raise RuntimeError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        if self.rank_aggr == "norm_2":
+            from pytools import norm_2
+            return norm_2
+        elif self.rank_aggr == "norm_inf":
+            from pytools import norm_inf
+            return norm_inf
+        elif self.rank_aggr == "sum":
+            return sum
+        elif self.rank_aggr == "min":
+            return min
+        elif self.rank_aggr == "max":
+            return max
+        else:
+            return None
+
+    def __call__(self):
+        """Return the requested quantity."""
+        raise NotImplementedError
+
+
+class ConservedQuantity(PhysicalQuantity):
+    """Logging support for conserved quantities (mass, energy, momentum)."""
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 unit: str = None, dim: int = None, name: str = None,
+                 rank_aggr: str = None):
+        if unit is None:
+            if quantity == "mass":
+                unit = "kg"
+            elif quantity == "energy":
+                unit = "J"
+            elif quantity == "momentum":
+                if dim is None:
+                    raise RuntimeError("Missing 'dim' parameter for dimensional "
+                                       f"ConservedQuantity '{quantity}'.")
+                unit = "kg*m/s"
```

Comment:
I don't think these are right.

- First, they're densities. (i.e. `/m^3`)
- Second, I'm not sure that "kg" is our fundamental mass unit. @MTCam?
- Third, it's really not the role of this code to have in-depth knowledge about physics. We should have a `FluidDynamics` object that knows this, the EOS, the dimensionality and such.

Reply:
Units used in the simulation are set by the user.  They are not inherent to the code, but the user's input data.   This is all done when the user sets the EOS, and the corresponding initial state.  For example, the user may:

```
eos = IdealSingleGas(gas_constant=8.314xxxxx)
```
This setting (which is the default) is consistent with kg (mass), Pascals (pressure), Kelvin (temp), etc.   Or the user can set:
```
eos = IdealSingleGas(gas_constant=.73xxxx)
```
Such a setting would be for users who provide initial state with lb, atm, Rankine.   

The user needs to make sure they specify the gas constant R, and initial state with consistent units.  CFD users are well-accustomed to this.  The built-in solution initializers and EOS are consistent with each other so that initial states (using built-in initializers) have the proper values consistent with the units.

If we have units specified anywhere in output data, they need to be settable by the user.  One way to make it automatically consistent is to just label it with the physical quantity (i.e. mass/volume, energy/volume, momentum/volume) for now.